### PR TITLE
Support sprockets 3

### PR DESCRIPTION
**Pull request by betesh**
Alternative to #111

### CODE REVIEW

**Comment by chussenot** [comment]
@kirs the pull request is good for you ? :cake: 

https://stackoverflow.com/questions/29617239/capistrano-3-sprockets-3-rails-4-2-1-wont-deploy/29636973#29636973

File: lib/capistrano/tasks/assets.rake [hunk 1]
```diff
@@ -62,15 +62,29 @@ namespace :deploy do
...
+    def with_backup_and_manifest_paths &block
+      backup_path = release_path.join('assets_manifest_backup')
+      manifest_path = test(:ls, manifest_path('.sprockets-manifest*'))
+      if manifest_path
+        manifest_path = capture(:ls, manifest_path('.sprockets-manifest*')).strip
+      else
+        manifest_path = capture(:ls, manifest_path('manifest*')).strip
+      end
+
+      yield manifest_path, backup_path
```
Comment (kirs) [inline lib/capistrano/tasks/assets.rake]: Do we really need `yield` here? I think we can just return these two variables.

File: lib/capistrano/tasks/assets.rake [hunk 2]
```diff
@@ -62,15 +62,29 @@ namespace :deploy do
...
+      release_path.join('public', fetch(:assets_prefix), filename)
+    end
+
+    def with_backup_and_manifest_paths &block
+      backup_path = release_path.join('assets_manifest_backup')
+      manifest_path = test(:ls, manifest_path('.sprockets-manifest*'))
+      if manifest_path
+        manifest_path = capture(:ls, manifest_path('.sprockets-manifest*')).strip
+      else
+        manifest_path = capture(:ls, manifest_path('manifest*')).strip
```
Comment (Tietew) [inline lib/capistrano/tasks/assets.rake]: This seems to break #91. Should be `'manifest*.*'` I think.

Comment (kirs) [inline lib/capistrano/tasks/assets.rake]: Yep!

File: lib/capistrano/tasks/assets.rake [hunk 3]
```diff
@@ -62,15 +62,29 @@ namespace :deploy do
       end
     end
 
+    def manifest_path(filename)
+      release_path.join('public', fetch(:assets_prefix), filename)
+    end
+
+    def with_backup_and_manifest_paths &block
```
Comment (chulkilee) [inline lib/capistrano/tasks/assets.rake]: As @kirs mentioned [below](https://github.com/capistrano/rails/pull/112/files#r28380633), why does it accept block instead of returning them?

Also missing parentheses around arguments.

**Comment by Tietew** [comment]
deploy works fine.
But deploy:rollback failed.

```
DEBUG [ccd2e772] Running /usr/bin/env ls /home/***/releases/20150415041452/public/assets/.sprockets-manifest* as ***@***.jp
DEBUG [ccd2e772] Command: cd /home/***/releases/20150415041452 && ( RBENV_ROOT=/usr/local/rbenv RBENV_VERSION=2.2.2 /usr/bin/env ls /home/***/releases/20150415041452/public/assets/.sprockets-manifest* )
DEBUG [ccd2e772]        /home/***/releases/20150415041452/public/assets/.sprockets-manifest-b93aabbf4844125525f01edf849a1740.json
DEBUG [ccd2e772] Finished in 0.016 seconds with exit status 0 (successful).
DEBUG [ba148f71] Running /usr/bin/env [[ -f /home/***/releases/20150415041452/assets_manifest_backup && -f /home/***/releases/20150415041452/public/assets/.sprockets-manifest-b93aabbf4844125525f01edf849a1740.json ]] as ***@***.jp
DEBUG [ba148f71] Command: [[ -f /home/***/releases/20150415041452/assets_manifest_backup && -f /home/***/releases/20150415041452/public/assets/.sprockets-manifest-b93aabbf4844125525f01edf849a1740.json ]]
DEBUG [ba148f71] Finished in 0.014 seconds with exit status 1 (failed).
WARN Rails assets manifest file (or backup file) not found.
(Backtrace restricted to imported tasks)
cap aborted!
SSHKit::Runner::ExecuteError: Exception while executing as ***@***.jp: Rails assets manifest file (or backup file) not found.
```

File: lib/capistrano/tasks/assets.rake [hunk 4]
```diff
@@ -62,15 +62,29 @@ namespace :deploy do
...
+      backup_path = release_path.join('assets_manifest_backup')
+      manifest_path = test(:ls, manifest_path('.sprockets-manifest*'))
```
Comment (chulkilee) [inline lib/capistrano/tasks/assets.rake]: Using the same name for method and local vars doesn't look good to me. In addition it is not path actually.

File: lib/capistrano/tasks/assets.rake [hunk 5]
```diff
@@ -78,15 +92,14 @@ namespace :deploy do
...
-          target = capture(:ls, release_path.join('public', fetch(:assets_prefix),
-                                                  'manifest*')).strip
-          if test "[[ -f #{source} && -f #{target} ]]"
-            execute :cp, source, target
-          else
-            msg = 'Rails assets manifest file (or backup file) not found.'
-            warn msg
-            fail Capistrano::FileNotFound, msg
+          with_backup_and_manifest_paths do |manifest_path, backup_path|
+            if test "[[ -f #{backup_path} && -f #{manifest_path} ]]"
```
Comment (Tietew) [inline lib/capistrano/tasks/assets.rake]: This `-f` will fail because PR #105 changes `assets_manifest_backup` to a directory.
If it's a directory, simply copying back `assets_manifest_backup/*` to `public/assets` seems to work fine.

Comment (Tietew) [inline lib/capistrano/tasks/assets.rake]: I think this is a bug of #105.
PR #84 includes fix of it.

File: lib/capistrano/tasks/assets.rake (hunk 1 above)
Comment (betesh) [inline lib/capistrano/tasks/assets.rake]: I don't think we need it--matter of personal taste really.  I think it looks cleaner but will happily change it if you feel strongly about it.

File: lib/capistrano/tasks/assets.rake (hunk 2 above)
Comment (betesh) [inline lib/capistrano/tasks/assets.rake]: fixed

File: lib/capistrano/tasks/assets.rake (hunk 4 above)
Comment (betesh) [inline lib/capistrano/tasks/assets.rake]: Please suggest a better name.  Which one does it make more sense to rename?

File: lib/capistrano/tasks/assets.rake (hunk 5 above)
Comment (betesh) [inline lib/capistrano/tasks/assets.rake]: Correct--I didn't change the functionality of the `test -f`.  That should be addressed as a separate issue.

File: lib/capistrano/tasks/assets.rake (hunk 4 above)
Comment (chulkilee) [inline lib/capistrano/tasks/assets.rake]: Just use it as if condition, not save it to var. Most capistrano code use test result directly in if condition, as you can see even in this file.

File: lib/capistrano/tasks/assets.rake (hunk 3 above)
Comment (chulkilee) [inline lib/capistrano/tasks/assets.rake]: There is `with` method to set env in target machine. This method name can be confusing without merit.

File: lib/capistrano/tasks/assets.rake (hunk 4 above)
Comment (betesh) [inline lib/capistrano/tasks/assets.rake]: Done.  But that doesn't address the use of `manifest` on the next 2 lines

Comment (chulkilee) [inline lib/capistrano/tasks/assets.rake]: Actually `manifest_path` method name is wrong since it is not returning manifest path.

I'd move the logic to detect actual manifest path to manifest_path method and drop `filename` arg from it. Then the method is responsible for detecting manifest path.

And then rename the method to `detect_manifest_path` and store the result to manifest_path variable to memorize in tasks calling it.

With this refactoring, using_manifest_and_backup_paths method will be just calling manifest_path, not including any logic around manifest path detection. Therefore I'd extract `asset_manifest_backup_path` method and let tasks to call the method and manifest_path directly, not via `using_manifest_and_backup_paths`.

File: lib/capistrano/tasks/assets.rake [hunk 6]
```diff
@@ -62,15 +62,28 @@ namespace :deploy do
       end
     end
 
+    def manifest_path(filename)
+      release_path.join('public', fetch(:assets_prefix), filename)
+    end
+
+    def manifest_and_backup_paths
```
Comment (chulkilee) [inline lib/capistrano/tasks/assets.rake]: The problem here is that `manifest_and_backup_paths` method has logic to determine manifest_path and that logic is totally unrelated to backup path. Do not put two different thing in one method.

Also extract a method to just get path for assets_manifest_backup.

Comment (kirs) [inline lib/capistrano/tasks/assets.rake]: :+1: for @chulkilee

Comment (betesh) [inline lib/capistrano/tasks/assets.rake]: @kirs I'm surprised you're seconding @chulkilee 

For some reason, I'm being held to a much higher standard here than the original developers of this gem.  I was the one to suggest some changes to make this code more maintainable, and then I went and did it.

The point of the `manifest_path` method was never to return the `manifest_path`, it was to reduce repetition, and I have asked @chulkilee to propose a more appropriate name for the method in light of that fact, and he has not made such a suggestion.  Instead, he proposed another name (not necessarily a better one IMO) for `backup_path`, a variable that I did not name.

As for @chulkilee's objection to doing 2 things in one method, neither of these 2 things was in a separate method at all before this PR--they were mixed into the code of each of the 2 tasks.  Moving them into a re-usable method is obviously an improvement.  Moving a one-liner (`backup_path = ...`) that is not duplicated anywhere into a separate method is not such an improvement--if someone submitted a PR with only that change, would you consider that enough of a change to be worth merging?

**Comment by chulkilee** [comment]
@Tietew  that is a bug in restore_manifest task. I created a separate PR - see #115

**Comment by chulkilee** [comment]
See #114 for minimal changes.

**Comment by betesh** [comment]
@kirs Feel free to close this in favor of #114

File: lib/capistrano/tasks/assets.rake (hunk 6 above)
Comment (chulkilee) [inline lib/capistrano/tasks/assets.rake]: (eek my previous comment is gone don't know what happened)

DRY isn't always good. There are bad DRY. Also DRY doesn't necessarily improve maintainability and readability.

Considering that capistrano tasks are pretty procedural, it would be better to explicitly call two methods, instead of adding a method to wrap two methods in one.

Also Capistrano has two block methods, `with` and `within_...` to wrap **host** environments. They must be block methods (SSHKit). So adding a block method to just wrap two methods can confuse readers.

I really appreciate your time and efforts on it. Like you, I wanted to keep the code quality of this gem pretty high, so that's why I jumped into your PR.

**Comment by i5okie** [comment]
Im not sure if I did this correctly. But I've installed capistrano-rails from chulkilee's repo (assuming with this fix). And I'm still getting the error..

```
SSHKit::Command::Failed: cp exit status: 1
cp stdout: Nothing written
cp stderr: cp: cannot stat ‘/app/releases/20150416230538/public/assets/manifest*’: No such file or directory

Tasks: TOP => deploy:assets:backup_manifest
(See full trace by running task with --trace)
The deploy has failed with an error: Exception while executing as deploy@11.22.33.44: cp exit status: 1
cp stdout: Nothing written
cp stderr: cp: cannot stat ‘/app/releases/20150416230538/public/assets/manifest*’: No such file or directory
```

I've tried, ```cap production deploy:cleanup``` and then deploy.. but still keep getting this error.

**Comment by chulkilee** [comment]
@i5okie please leave comments on #114 if you use that code. Also make sure you use the correct branch - it's not in master.

**Comment by kirs** [comment]
Fixed in master. Thanks @chulkilee!

**Comment by kirs** [comment]
@betesh thank you for your work!
